### PR TITLE
Added Updates dashboard widget

### PR DESCRIPTION
**Pull request by konadrian**
Added Updates dashboard widget
WidgetUpdates

### CODE REVIEW

**Comment by mariuszkrzaczkowski** [comment]
## Migration Recommended

It appears you are still using the CoPilot OAuth App - we recommend you [migrate to the GitHub App](https://github.com/apps/blackduck-copilot/installations/new/permissions?suggested_target_id=10201412&repository_ids[]=28060246)

**Comment by mariuszkrzaczkowski** [comment]
## Black Duck Security Report

Merging #11180 into developer will not change security risk.

[Click here to see full report](https://copilot.blackducksoftware.com/github/repos/YetiForceCompany/YetiForceCRM/pull-requests/11180)

File: layouts/basic/modules/Vtiger/dashboards/Updates.tpl [hunk 1]
```diff
@@ -0,0 +1,42 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 3.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
+{strip}
+	{assign var=ACCESSIBLE_USERS value=\App\Fields\Owner::getInstance()->getAccessibleUsers()}
```
[the remark this answers is not in the document]
Comment (konadrian) [reply layouts/basic/modules/Vtiger/dashboards/Updates.tpl]: It is use in SelectAccessibleTemplate.tpl

File: layouts/basic/modules/Vtiger/dashboards/Updates.tpl [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	{assign var=ACCESSIBLE_GROUPS value=\App\Fields\Owner::getInstance()->getAccessibleGroups()}
```
[the remark this answers is not in the document]
Comment (konadrian) [reply layouts/basic/modules/Vtiger/dashboards/Updates.tpl]: It is use in SelectAccessibleTemplate.tpl

File: app/Module.php [hunk 3]
```diff
@@ -292,6 +292,49 @@ public static function changeType(string $moduleName, int $type)
...
+	 */
+	public static function getPermittedUserModuleModules(): array
+	{
+		$userId = User::getCurrentUserId();
+		if (Cache::staticHas('getPermittedUserModuleModules', $userId)) {
+			return Cache::staticGet('getPermittedUserModuleModules', $userId);
+		}
+	$query = (new Db\Query())->from('vtiger_tab')
+			->where(['<>', 'presence', 1]);
+		$query->andWhere(['not in', 'name', ['Users', 'Dashboard', 'Home', 'Import', 'ModComments', 'PriceBooks', 'CallHistory', 'OSSMailView', 'SMSNotifier', 'RecycleBin', 'Modtracker']]);
```
Comment (rskrzypczak) [inline app/Module.php]: ? remove

File: app/Module.php [hunk 4]
```diff
@@ -292,6 +292,49 @@ public static function changeType(string $moduleName, int $type)
 			UserPrivilegesFile::recalculateAll();
 		}
 	}
+
+	/**
+	 * Function get user permitted module modules
+	 *
+	 * @return array
+	 */
+	public static function getPermittedUserModuleModules(): array
```
Comment (rskrzypczak) [inline app/Module.php]: ? remove

File: app/Module.php [hunk 5]
```diff
@@ -292,6 +292,49 @@ public static function changeType(string $moduleName, int $type)
...
+		Cache::staticSave('getPermittedUserModuleModules', $userId, $permittedModules);
+		return $permittedModules;
+	}
+
+	/**
+	 * Function get user permitted module ids
+	 *
+	 * @return array
+	 */
+	public static function getPermittedUserModulesId():array
```
Comment (rskrzypczak) [inline app/Module.php]: ?

File: layouts/basic/modules/Vtiger/dashboards/UpdatesContents.tpl [hunk 6]
```diff
@@ -0,0 +1,127 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 3.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
+{strip}
+	<div class="tpl-Base-dashboards-UpdateContents">
+		{if $UPDATES neq false}
+			{foreach item=UPDATE_ROW from=$UPDATES}
+				{assign var=MODEL_NAME value=get_class($UPDATE_ROW)}
```
Comment (rskrzypczak) [inline layouts/basic/modules/Vtiger/dashboards/UpdatesContents.tpl]: ?

File: layouts/basic/modules/Vtiger/dashboards/UpdatesContents.tpl [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+									<strong>{$USER->getName()}&nbsp;</strong>{\App\Language::translate($UPDATE_ROW->getStatusLabel(), 'ModTracker')}
+									<a class="u-cursor-pointer" {if stripos($DETAILVIEW_URL, 'javascript:')===0} onclick='{$DETAILVIEW_URL|substr:strlen("javascript:")}' {else} onclick='window.location.href = "{$DETAILVIEW_URL}"' {/if}>
+										&nbsp;{$PARENT->getName()}
+									</a>
+								</div>
+							{/if}
+						</div>
+					</div>
+				{/if}
+				{else if $MODEL_NAME == 'ModComments_Record_Model'}
```
Comment (rskrzypczak) [inline layouts/basic/modules/Vtiger/dashboards/UpdatesContents.tpl]: ?

File: modules/Home/models/Module.php [hunk 8]
```diff
@@ -296,4 +296,90 @@ public function getHistory($pagingModel, $type = false)
...
+				$modules[] = $moduleName;
+			}
+		}
+		$widgetData['selectedModules'];
+		$query = new \App\Db\Query();
+		$query->select(['vtiger_modtracker_basic.*'])
+			->from('vtiger_modtracker_basic')
+			->innerJoin('vtiger_crmentity', 'vtiger_modtracker_basic.crmid = vtiger_crmentity.crmid')
+			->where(['vtiger_crmentity.deleted' => 0, 'module' => $modules, 'vtiger_modtracker_basic.status' => $widgetData['selectedTrackerActions']]);
+		$query->andWhere(['between', 'changedon', $dateRange[0] . ' 00:00:00', $dateRange[1] . ' 23:59:59']);
```
Comment (rskrzypczak) [inline modules/Home/models/Module.php]: dateRange  db format

File: modules/Home/models/Module.php [hunk 9]
```diff
@@ -296,4 +296,90 @@ public function getHistory($pagingModel, $type = false)
...
+	/**
+	 *	Function get available modules for user
+	 */
+	public static function getModulesForUpdateWidgets(){
+		$userId = \App\User::getCurrentUserId();
+		if (\App\Cache::staticHas('getModulesForUpdateWidgets', $userId)) {
+			return \App\Cache::staticGet('getModulesForUpdateWidgets', $userId);
+		}
+		$permittedModules = \App\Module::getPermittedUserModuleModules();
+		$menu = \Vtiger_Menu_Model::getAll(true);
```
Comment (rskrzypczak) [inline modules/Home/models/Module.php]: ?

File: modules/Vtiger/views/UpdatesWidgetConfig.php [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+ * @author    Adrian Koń <a.kon@yetiforce.com>
+ */
+class Vtiger_UpdatesWidgetConfig_View extends \App\Controller\Modal
+{
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function checkPermission(\App\Request $request)
+		{
```
Comment (rskrzypczak) [inline modules/Vtiger/views/UpdatesWidgetConfig.php]: phpcs

File: modules/Vtiger/views/UpdatesWidgetConfig.php [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+			$this->widget = Vtiger_Widget_Model::getInstanceWithWidgetId($request->getInteger('widgetId'), \App\User::getCurrentUserId());
+			if (!\App\Privilege::isPermitted($request->getModule()) || $this->widget->get('userid') !== \App\User::getCurrentUserId()) {
```
Comment (rskrzypczak) [inline modules/Vtiger/views/UpdatesWidgetConfig.php]: ?

**Pull request closed without merging**